pkg/controllers/dashboard/cspadaptercharts/controller.go

```diff
@@ -0,0 +1,51 @@
+// Package cspadaptercharts creates the controllers necessary for automatically upgrading the csp adapter chart
+package cspadaptercharts
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/catalogv2/system"
+	"github.com/rancher/rancher/pkg/managedcharts/cspadapter"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/rancher/pkg/wrangler"
+)
+
+func Register(ctx context.Context, wContext *wrangler.Context) error {
+	h := &handler{
+		manager:     wContext.SystemChartsManager,
+		adapterUtil: cspadapter.NewChartUtil(wContext.RESTClientGetter),
+	}
+	wContext.Mgmt.Setting().OnChange(ctx, "csp-adapter-upgrade", h.onSetting)
+
+	return nil
+}
+
+type handler struct {
+	sync.Mutex
+	manager     *system.Manager
+	adapterUtil *cspadapter.ChartUtil
+}
+
+func (h *handler) onSetting(key string, setting *v3.Setting) (*v3.Setting, error) {
+	if setting == nil {
+		return nil, nil
+	}
+	if setting.Name != settings.CSPAdapterMinVersion.Name {
+		// we only need to check for changes to the CSPAdapterMinVersion setting, ignore any others
+		return setting, nil
+	}
+	adapterRelease, err := h.adapterUtil.GetRelease()
+	if err != nil {
+		if errors.Is(err, cspadapter.ErrNotFound) {
+			// If the adapter isn't installed, don't attempt an upgrade
+			return setting, nil
+		}
+		// if we can't determine the status of the adapter, stop here and attempt to re-evaluate later
+		return setting, err
+	}
+	err = h.manager.Ensure(cspadapter.ChartNamespace, adapterRelease.Name, settings.CSPAdapterMinVersion.Get(), nil, true)
```

Comment:
What if the current version is not equal to but greater than the minimum required? Would Rancher downgrade it?

Reply:
This is getting more into the functionality of the Ensure method that we are using here, since that is what decides if the app should be installed/upgraded.

That being said, I believe that it will only upgrade if the desired version (min) is higher than the actual version. See [here](https://github.com/rancher/rancher/blob/952e27f9dd9bf9ebfb2226795fd2d34febcb0fd4/pkg/catalogv2/system/system.go#L409-L431) for the relevant code.